pkg/virt-handler/isolation/isolation.go

```diff
@@ -45,6 +45,9 @@ type PodIsolationDetector interface {
 	// It returns an IsolationResult containing all isolation information
 	Detect(vm *v1.VirtualMachineInstance) (*IsolationResult, error)
 
+	// Detects the PID of the VMI with the given UID
+	GetPid(vmiUid string) (int, error)
```

Comment:
If you use on the rest-api namespace/name, you can look up the vmi from the informer cache and use the normal `Detect` flow. I would prefer that, since things have the one canonical entrypoint, no matter if on the api or in the controller loop.

Reply:
@rmohr not sure I'm following. by `I would prefer that, since things have the one canonical entrypoint, no matter if on the api or in the controller loop.` you mean that the res-api in virt-handler would match the one of virt-api? if so, then yeah, I agree that it would be better to change the rest-api to namespace/name, but let's do it as a separate PR

and regarding the `Detect` flow, the thing is that I don't need the slice part. that's why I separated out the retrieval of pid. I don't mind to call `Detect` with an option to skip retrieving the slice if you think it's better

Reply:
> but let's do it as a separate PR

why?

Reply:
> and regarding the `Detect` flow, the thing is that I don't need the slice part. that's why I separated out the retrieval of pid. I don't mind to call `Detect` with an option to skip retrieving the slice if you think it's better

Detecting the slice is cheap and we do it all the time in the controller func, I don't think it matters. You just open a pseudo-file in proc.

Reply:
> > but let's do it as a separate PR
> 
> why?

this PR in its current state already includes:
1. Opening the console through virt-handler rather than virt-api executing on virt-launcher
2. Dropping sock-connector
3. Dropping pods/exe priviledges from virt-api
4. Improving copying from/to/between websockets
5. Some refactoring
... and became XXL.
so I think that if we find that it meets the requirements and is stable and not negatively affects the existing code, we should merge it and beautify the code separately. practically, that minimizes rebases on my side and for others.

Reply:
> > and regarding the `Detect` flow, the thing is that I don't need the slice part. that's why I separated out the retrieval of pid. I don't mind to call `Detect` with an option to skip retrieving the slice if you think it's better
> 
> Detecting the slice is cheap and we do it all the time in the controller func, I don't think it matters. You just open a pseudo-file in proc.

honestly, I would prefer to split the code into smaller reusable pieces that can be invoked independently, depending on the desired output. but let's not argue about it, I will call `Detect` if you like

Reply:
> so I think that if we find that it meets the requirements and is stable and not negatively affects the existing code, we should merge it and beautify the code separately. practically, that minimizes rebases on my side and for others.

I don't see how this has a smaller or bigger effect for others. It is a question of luck if you have right now a PR open which will need a rebase after this one is merged, or if you have one open a few days later when we merge the PR.

This change will have immediate effect after the PR is merged. If you don't manage to prepare and merge the follow-up PR until we do a release, virt-handler will strictly speaking have to support both apis. the one with the UID and the one with namespace/name.

Every commit on master is a release candidate and therefore we extra have all testing possibilities on the PR itself.

Reply:
> > so I think that if we find that it meets the requirements and is stable and not negatively affects the existing code, we should merge it and beautify the code separately. practically, that minimizes rebases on my side and for others.
> 
> I don't see how this has a smaller or bigger effect for others. It is a question of luck if you have right now a PR open which will need a rebase after this one is merged, or if you have one open a few days later when we merge the PR.
> 
> This change will have immediate effect after the PR is merged. If you don't manage to prepare and merge the follow-up PR until we do a release, virt-handler will strictly speaking have to support both apis. the one with the UID and the one with namespace/name. Otherwise seamless updates would not work for console and vnc.
> 
> Every commit on master is a release candidate and therefore we extra have all testing possibilities on the PR itself.

I see, so can we drop pods/exec from virt-api? or drop sock-connector? because then I wonder what will happen when old virt-api talks with new virt-handler that launches virt-launcher with no sock-connector?

Reply:
> If you don't manage to prepare and merge the follow-up PR until we do a release, virt-handler will strictly speaking have to support both apis.

oh yeah... I see the risk here. We have to make a decision about the rest endpoint and execute on it in this PR. I think namespace/name with the informer lookup is the way forward. 

>  see, so can we drop pods/exec from virt-api? or drop sock-connector? because then I wonder what will happen when old virt-api talks with new virt-handler that launches virt-launcher with no sock-connector?

Here's what I suggest. 

* keep socat and sock-connector for now. We can safely remove it the release after next. 
* don't change the kubevirt-private directory tree. Otherwise we risk breaking console/vnc access with existing running VMIs during update. We're pretty much locked in to that directory structure unless we want to go through the hassle of attempting to support two paths... probably not worth it 
* Make the virt-handler vnc/console rest api use namespace/name in the url, then translate the namespace/name to the right kubevirt-private directory tree (which uses uuid) by doing a lookup in the vmi informer cache.

sound reasonable?

Reply:
> If you don't manage to prepare and merge the follow-up PR until we do a release, virt-handler will strictly speaking have to support both apis.
> 
> oh yeah... I see the risk here. We have to make a decision about the rest endpoint and execute on it in this PR. I think namespace/name with the informer lookup is the way forward.

@fabiand , please correct me if I'm wrong here:
IIUC the console that virt-api exposes is intended for installations. For ordinary use by users, VDI-like use, we count on other guest-level tools like RDP.
So I see the that updating to new versions on virt-api/virt-hander that cannot interact with existing virt-launcher pods is something we should avoid. However, I think it is ok then to assume that console connections would be terminated on update and that users would reinstantiate the connection. so I don't see the risk yet, can you please clarify it?

> 
> see, so can we drop pods/exec from virt-api? or drop sock-connector? because then I wonder what will happen when old virt-api talks with new virt-handler that launches virt-launcher with no sock-connector?
> 
> Here's what I suggest.
> 
> keep socat and sock-connector for now. We can safely remove it the release after next.

but do I understand it correctly that we re required to support partial updates where for instance virt-api is not updated and virt-handler is updated? if so, how can we ever change the API between those components non-backward compatible way?

> don't change the kubevirt-private directory tree. Otherwise we risk breaking console/vnc access with existing running VMIs during update. We're pretty much locked in to that directory structure unless we want to go through the hassle of attempting to support two paths... probably not worth it
> Make the virt-handler vnc/console rest api use namespace/name in the url, then translate the namespace/name to the right kubevirt-private directory tree (which uses uuid) by doing a lookup in the vmi informer cache.
> 
> sound reasonable?

yeah, ok.

Reply:
> I see why updating to new versions on virt-api/virt-hander that cannot interact with existing virt-launcher pods is something we should avoid. However, I think it is ok to assume that console connections are terminated on update and that users would reinstantiate the connection. so I don't see the risk yet, can you please clarify it?

General things: We first update all virt-handlers, then virt-controller and finally virt-api. This means that if we have virt-api to virt-handler connections, we need to be backward compatible on the virt-handler api for at least one release, since it is not really predictable how long it takes to update all virt-handlers (or if we even end up with a mixed state for a long time).

Witht he socat approach we actually have a depencency between virt-api, virt-controller and virt-launcher. Now, once virt-controller is updated, it will create new launcher pods, but virt-api will still try to connect the old way. There may be temporary issues which prevent the new virt-api to become ready. until this is the case, the old virt-api is still in place and can't establish new connections for a potentially long time.

> but do I understand it correctly that we re required to support partial updates where for instance virt-api is not updated and virt-handler is updated? if so, how can we ever change the API between those components non-backward compatible way?

Right now we try to support n-1 to n updates. Maybe n-2 to n soon. So we just have to keep internal things for one additional release.

Reply:
> I see, so can we drop pods/exec from virt-api? or drop sock-connector? because then I wonder what will happen when old virt-api talks with new virt-handler that launches virt-launcher with no sock-connector?

You are right, we have to keep it for at least one release. Nit: virt-controller lauchnes virt-launcher, not virt-handler.

Reply:
> I see why updating to new versions on virt-api/virt-hander that cannot interact with existing virt-launcher pods is something we should avoid. However, I think it is ok to assume that console connections are terminated on update and that users would reinstantiate the connection. so I don't see the risk yet, can you please clarify it?
> 
> General things: We first update all virt-handlers, then virt-controller and finally virt-api. This means that if we have virt-api to virt-handler connections, we need to be backward compatible on the virt-handler api for at least one release, since it is not really predictable how long it takes to update all virt-handlers (or if we even end up with a mixed state for a long time).
> Witht he socat approach we actually have a depencency between virt-api, virt-controller and virt-launcher. Now, once virt-controller is updated, it will create new launcher pods, but virt-api will still try to connect the old way. There may be temporary issues which prevent the new virt-api to become ready. until this is the case, the old virt-api is still in place and can't establish new connections for a potentially long time.
>

ok, so by "seamless upgrade" in regards to the console connections we actually mean that on upgrade existing console connections could be terminated (because they go through at least virt-api that would restart) but it should be possible to reinitiate them "immediately", i.e., also during "partial upgrade", where only  some of the components in the cluster were upgraded, right?

Is that a clear requirement that we got from a stakeholder? I'm asking because under the assumption the console that is provided by kubevirt is only used for installations and one-time configuration, we may be able to relax this requirement..  but I guess that the console is not expect to change that often so not sure it would worth the time..

> but do I understand it correctly that we re required to support partial updates where for instance virt-api is not updated and virt-handler is updated? if so, how can we ever change the API between those components non-backward compatible way?
> 
> Right now we try to support n-1 to n updates. Maybe n-2 to n soon. So we just have to keep internal things for one additional release.

Alright, that makes sense.
So we'll keep `socat` and `sock-connector` for now and proceed as proposed by @davidvossel

Reply:
> k, so by "seamless upgrade" in regards to the console connections we actually mean that on upgrade existing console connections could be terminated (because they go through at least virt-api that would restart) but it should be possible to reinitiate them "immediately", i.e., also during "partial upgrade", where only some of the components in the cluster were upgraded, right?


correct. the expectation for now is that console/vnc connections will be broken during the upgrade.